### PR TITLE
Security experiments

### PR DESCRIPTION
**Pull request by pliniosilveira**

### CODE REVIEW

File: include/config.py [hunk 1]
```diff
@@ -102,7 +102,7 @@
 # (b bit of address = logN, N is the number of nodes)
 
 # Change here - Specify the Node parameters here
-num_tile_compute =7 # number of tiles mapped by dnn (leaving input and output tiles)
+num_tile_compute = 2 # number of tiles mapped by dnn (leaving input and output tiles)
```
Comment (pliniosilveira) [inline include/config.py]:
```suggestion
num_tile_compute = 7 # number of tiles mapped by dnn (leaving input and output tiles)
```

Just to reduce changes.

File: src/dpe.py [hunk 2]
```diff
@@ -79,6 +108,11 @@ def run(self, net):
 
         if not os.path.exists(tracedir):
             os.makedirs(tracedir)
+
+
+            
+
+
```
Comment (pliniosilveira) [inline src/dpe.py]: Remove blank lines

File: test/utils/generate-py.sh [hunk 3]
```diff
@@ -1,7 +1,7 @@
 # Copy this file to compiler test folder where the .npy files are generated.
 # Update SIMULATOR_PATH value and execute it.
 
-SIMULATOR_PATH=" " # simulator root path
+SIMULATOR_PATH="/home/marte/Desktop/git_fork/puma-simulator" # simulator root path
```
Comment (pliniosilveira) [inline test/utils/generate-py.sh]: Do not commit personal paths.

```suggestion
SIMULATOR_PATH=" " # simulator root path
```

File: test/utils/populate.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 import os
 import numpy as np
 
-SIMULATOR_PATH=" "
+SIMULATOR_PATH="/home/marte/Desktop/puma-simulator"
```
Comment (pliniosilveira) [inline test/utils/populate.py]: Do not commit personal paths.

**Review by pliniosilveira — COMMENTED** [review]
Please remove binary files: 
images/puma.png

File: src/dpe.py (hunk 2 above)
Comment (yliborio) [reply src/dpe.py]: resolved

File: test/utils/populate.py (hunk 4 above)
Comment (yliborio) [reply test/utils/populate.py]: resolved

File: Security/how_to_run_with_security.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+- [How to run](#how-to-run)
+
+## Alterations Made
+
+Two different solutions for security were implemented, first a MAC(Message Authentication Code) was implemented to validate models received by the architecture and secondly a model encryption approach was developed to allow the model to be securely sent to the simulator.
+
+For the security module, we implemented two interfaces that allow for different implementations of cryptography and authentication methods. We also made one implementation for each interface, the first one was for authentication and uses a SHA256 hash to generate a hash of the model and the input and then we used the Fernet python library to encrypt the hash generating a MAC. The Fernet library was also used for the encryption/decryption of models. We also created a factory class in order to make adding new security models easier. Other alterations made to run models securely were in the source file dpe.py in which two parameters -a and -c that allow the simulator to run authenticated models “-a” or encrypted models “-c” were added.
+
+  
+![UML Diagram](/images/puma.png)
```
Comment (pliniosilveira) [inline Security/how_to_run_with_security.md]:
```suggestion
![image](https://user-images.githubusercontent.com/2287889/75450667-fe2df500-594d-11ea-9f75-e55f2af90448.png)
```

**Pull request merged**